swift-sdk/Internal/Promise.swift

```diff
@@ -0,0 +1,118 @@
+//
+//  Promise.swift
+//  swift-sdk
+//
+//  Created by Tapash Majumder on 10/26/18.
+//  Copyright © 2018 Iterable. All rights reserved.
+//
+
+import Foundation
+
+enum Result<Value, ErrorType> {
+    case value(Value)
+    case error(ErrorType)
+}
+
+// This has only two public methods
+// either there is a success with result
+// or there is a failure with error
+// There is no way to set value a result in this class.
+class Future<Value, ErrorType> {
+    var onSuccess: ((Value) -> Void)? = nil {
```

Comment:
Isn't this usually a function that returns another `Future` so you can chain these?
```
future.flatMap({...}).onSuccess({...}).onFailure({...})
```